src/Hamiltonians/geometry.jl

```diff
@@ -1,256 +1,227 @@
 """
-    abstract type LatticeGeometry{D}
+    Geometry(dims::NTuple{D,Int}, fold::NTuple{D,Bool})
 
-A `LatticeGeometry` controls which sites in an [`AbstractFockAddress`](@ref) are considered
-to be neighbours.
+Represents a `D`-dimensional grid. Used to convert between cartesian vector indices (tuples
+or `SVector`s) and linear indices (integers). When indexed with vectors, it folds them back
+into the grid if the out-of-bounds dimension is periodic and 0 otherwise (see example
+below).
 
-Currently only supported by [`HubbardRealSpace`](@ref).
+* `dims` controls the size of the grid in each dimension.
+* `fold` controls whether the boundaries in each dimension are periodic (or folded in the
+  case of momentum space).
 
-# Available implementations
+```julia
+julia> geo = Geometry((2,3), (true,false))
+Geometry{2}((2, 3), (true, false))
 
-* [`PeriodicBoundaries`](@ref)
-* [`HardwallBoundaries`](@ref)
-* [`LadderBoundaries`](@ref)
+julia> geo[1]
+(1, 1)
 
-# Interface to implement
+julia> geo[2]
+(2, 1)
 
-* `Base.size`: return the lattice size.
-* [`neighbour_site(::LatticeGeometry, ::Int, ::Int)`](@ref)
-* [`num_dimensions(::LatticeGeometry)`](@ref)
-* [`num_neighbours(::LatticeGeometry)`](@ref)
-"""
-abstract type LatticeGeometry{D} end
+julia> geo[3]
+(1, 2)
 
-function Base.show(io::IO, geom::LatticeGeometry)
-    print(io, nameof(typeof(geom)), size(geom))
-end
+julia> geo[(1,2)]
+3
 
-"""
-    onr(add::AbstractFockAddress, geom::LatticeGeometry)
-Returns the occupation number representation of a Fock state address as an `SArray`
-with the shape of the lattice geometry `geom`. For composite addresses, a tuple
-of `onr`s is returned.
-"""
-function BitStringAddresses.onr(add, geom::LatticeGeometry)
-    return reshape(onr(add), size(geom))
-end
-function BitStringAddresses.onr(add::CompositeFS, geom::LatticeGeometry)
-    return map(fs -> onr(fs, geom), add.components)
-end
+julia> geo[(3,2)] # 3 is folded back into 1
+3
 
-"""
-    neighbour_site(geom::LatticeGeometry, site, i)
+julia> geo[(3,3)]
+5
 
-Find the `i`-th neighbour of `site` in the geometry. If the move is illegal, return 0.
-"""
-neighbour_site
+julia> geo[(3,4)] # returns 0 if out of bounds
+0
 
+```
 """
-    num_dimensions(geom::LatticeGeometry)
+struct Geometry{D,Dims,Fold}
```

Comment:
I understand that `Geometry` is now a concrete type whereas the old `LatticeGeometry` was an abstract type. Nevertheless, I think I prefer the old name because it is more specific and intuitive. So my suggestion would be to rename `Geometry` to `LatticeGeometry`.

Reply:
Or maybe `CubicLattice`, as we may want to add other lattices (hexagonal, kagome, etc. ) later.